superset/assets/src/dashboard/util/activeDashboardFilters.js

```diff
@@ -16,35 +16,122 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+import { isEmpty } from 'lodash';
+
+import {
+  getChartIdAndColumnFromFilterKey,
+  getDashboardFilterKey,
+} from './getDashboardFilterKey';
+import { CHART_TYPE } from '../util/componentTypes';
+
+let allFilterBoxChartIds = [];
 let activeFilters = {};
+let appliedFilterValuesByChart = {};
+let allComponents = {};
 
+// output: { [id_column]: { values, scope } }
 export function getActiveFilters() {
   return activeFilters;
 }
 
-// non-empty filters from dashboardFilters,
-// this function does not take into account: filter immune or filter scope settings
-export function buildActiveFilters(allDashboardFilters = {}) {
-  activeFilters = Object.values(allDashboardFilters).reduce(
-    (result, filter) => {
-      const { chartId, columns } = filter;
-
-      Object.keys(columns).forEach(key => {
-        if (
-          Array.isArray(columns[key])
-            ? columns[key].length
-            : columns[key] !== undefined
-        ) {
-          /* eslint-disable no-param-reassign */
-          result[chartId] = {
-            ...result[chartId],
-            [key]: columns[key],
+// currently filter_box is a chart,
+// when selecting filter scopes, they have to be out pulled out in a few places.
+// after we make filter_box a dashboard build-in component, will not need this check anymore.
+export function isFilterBox(chartId) {
+  return allFilterBoxChartIds.includes(chartId);
+}
+
+// this function is to find all filter values applied to a chart,
+// it goes through all active filters and their scopes.
+// return: { [column]: array of selected values }
+export function getAppliedFilterValues(chartId) {
+  if (!(chartId in appliedFilterValuesByChart)) {
+    appliedFilterValuesByChart[chartId] = Object.entries(activeFilters).reduce(
```

Comment:
Might be able to use `lodash/fp` to improve the readability of this logic.

Reply:
yay!